From 30bac415ce1da0e17515e83bc102e0b5e38f9a7c Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:01:51 -0400
Subject: [usb] ehci: No SSPLIT allowed in uframe 7

Message-id: <1410887463-20674-40-git-send-email-dzickus@redhat.com>
Patchwork-id: 95577
O-Subject: [RHEL7 PATCH 039/591] USB: EHCI: No SSPLIT allowed in uframe 7
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 8c05dc598e5bc0eb33791de23157cf1e47cb580e
Author: Alan Stern <stern@rowland.harvard.edu>
Date:   Fri Oct 11 11:28:31 2013 -0400

    USB: EHCI: No SSPLIT allowed in uframe 7

    The scheduling code in ehci-hcd contains an error.  For full-speed
    isochronous-OUT transfers, the EHCI spec forbids scheduling
    Start-Split transactions in H-microframe 7, but the driver allows it
    anyway.  This patch adds a check to prevent it.

    Signed-off-by: Alan Stern <stern@rowland.harvard.edu>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/host/ehci-sched.c b/drivers/usb/host/ehci-sched.c
index 4b0903c..1fc2bef 100644
--- a/drivers/usb/host/ehci-sched.c
+++ b/drivers/usb/host/ehci-sched.c
@@ -1286,6 +1286,10 @@ sitd_slot_ok (
 
  mask = stream->raw_mask << (uframe & 7);
 
+ /* for OUT, don't wrap SSPLIT into H-microframe 7 */
+ if (((stream->raw_mask & 0xff) << (uframe & 7)) >= (1 << 7))
+  return 0;
+
  /* for IN, don't wrap CSPLIT into the next frame */
  if (mask & ~0xffff)
   return 0;
-- 
1.7.1